### ISSUE TITLE
Add bower support

### ISSUE THREAD

**Issue by sbordet** [cometd/cometd#545]
See https://github.com/cometd/cometd/pull/18

**Comment by lordkada** [comment]
Hi @sbordet, FYI:

To build the distribution js (jquery) I use a different approach from the one depicted by PR #18: essentially I just clone the cometd repo and concatenate all the js files (without compiling the java demo...)
 
You can checkout it yourself by looking at the build.sh here https://github.com/lordkada/cometd-javascript-bower

Ciao,
CA

**Comment by chchrist** [comment]
I suggest we move all the JS to its own project, with a proper package.json and example, no maven (which makes no sense for a front end developer). The way it is structured now is slowing down adoption...

**Comment by sbordet** [comment]
All right, so I went and had a bower crash course. 

The idea would be to deploy CometD to the bower repository, so that people will use bower to install CometD.

The issue I currently have is that CometD binds with several toolkits, in particular Dojo, jQuery and Angular.

So to create a project, one would do:

```bash
$ bower install cometd-angular
```

The command above will install CometD with the Angular binding.

Similarly, for use with the jQuery binding, one would do:

```bash
$ bower install cometd-jquery
```

and similar for the Dojo binding.

The issue is that I cannot have 3 different bower files at the root of the project, like [the instructions say](http://bower.io/docs/creating-packages/#register).

Do you guys know if it is possible to have a single `bower.json` with multiple names in it, or if it is possible to have the `bower.json` files not at the root of the project ?

**Comment by lordkada** [comment]
@sbordet I don't think it is possible to do what you are asking for... unfortunately I think you need 3 different packages... (or 4 if you want to move out the "common js", i.e. cometd, cometd-jquery, cometd-dojo and cometd-angular)

**Comment by sbordet** [comment]
So can I ask what use is Bower to you guys ?

From what I understand it is only a dependency manager.

If the problem is to have a quick startup of the project, then Maven archetypes give you so much more than Bower.

Bower-style:
 * Install node
 * Install npm
 * Install bower
 * `bower install cometd-jquery`
 * Create `index.html`
 * Manually add all JavaScript dependencies - no errors please
 * Manually write `application.js` - look at other examples
 * Install Java
 * Manually write server-side hello world Java class
 * Manually write some `pom.xml` or `gradle.build` build files to create the *.war file
 * Build & Run

Maven style:
 * Install Java
 * Install Maven
 * `mvn archetype:generate -DarchetypeCatalog=http://cometd.org`
 * Build & Run

To me, Maven wins hands down in simplicity.

Since I fail to see how Bower is useful here, can you please enlighten me ? 
I am obviously missing something.

Thanks !

**Comment by lordkada** [comment]
Pure client-side applications have a totally different approach and usually you don't go through server-side (or hybrid) technologies like Java, Rails or whatever.

Typically you deal with a bunch of CSS, HTML, JS (or other transpilled languages). Of course is not that basic, but it is just to give you the idea of what is involved.

An example of framework used in these scenarios is [Ember.js](http://emberjs.com)

**Comment by sbordet** [comment]
Still I don't get how you can write a pure client-side application with CometD. You will need the server-side.

As far as I understand, also Ember.js needs a server side, just that it's written in JavaScript and run by Node.

I'm still missing what benefit would Bower bring to the table.

Other use cases that may penetrate my think skull ?

How do **you** use Bower ?

**Comment by lordkada** [comment]
Yes, CometD needs a java-server, of course... but that doesn't mean that you're going to use Java to build your UI...(I dropped this approach several years ago; front and back ends may have totally different needs in terms of environment, architecture, technology, skillset and so on...). People with back-end skills continue to develop using Java/Maven.. Front-end developers just use different stuffs...

No... the Ember.js output apps are static .js/.css files (in my own case I serve them just pushing to  AWS/S3/CloudFront)... Generally speaking Node.js is used to serve it locally (development) or to produce the distribution files and then to push them to your CDN (e.g. http://ember-cli.com/ember-cli-deploy/)

When you develop this way,  **bower (or npm) is your dependency manager**, because on the front-end side you don't have anything else than javascript (you use node.js to mangle your files or serve locally... but sometimes it happened that I simply used nginx to run/develop my app on my computer!)

This is an excerpt from one of my bowers:

```
{
  "name": "<omissis>",
  "dependencies": {
    "async": "~1.4.2",
    "bootstrap": "~3.3.5",
    "cometd-javascript-bower": "git@github.com:lordkada/cometd-javascript-bower.git#v3.0.5",
    "ember": "1.13.10",
    "ember-cli-shims": "ember-cli/ember-cli-shims#0.0.6",
    "ember-cli-test-loader": "ember-cli-test-loader#0.1.3",
    "ember-load-initializers": "ember-cli/ember-load-initializers#0.1.5",
    "ember-qunit": "0.4.9",
    "ember-qunit-notifications": "0.0.7",
    "ember-resolver": "~0.1.18",
    "fastclick": "~1.0.6",
    "jquery": "~2.1.4",
    "jquery-validation": "~1.13",
    "jquery.easy-pie-chart": "~2.1.6",
    "loader.js": "ember-cli/loader.js#3.2.1",
    "material-design-iconic-font": "~2.1.2",
    "material-shadows": "~2.0.1",
    "moment": "~2.10.6",
    "qunit": "~1.18.0",
    "roboto-webfont-bower": "roboto-webfont#~0.1.1",
    "<omissis>-js-client-library": "git@bitbucket.<omissis>/js-client-library.git#v0.0.1",
    "<omissis>-js-utils-library": "git@bitbucket.<omissis>/js-utils.git#v0.0.2",
    "showdown": "~1.2.3",
    "magnific-popup": "~1.0.0"
  }
}
```

**Comment by sbordet** [comment]
Carlo,

sorry but I still don't understand.

If you had CometD under Bower, how would you use it ?
Sure you put it in your bower.json, but then what ?

I can only think of building some static page without being able to test anything... it'll be of little use to me.

You may open the page in the browser, but it won't work at all (as soon as your JavaScript hits the CometD APIs).

```bash
$ bower install cometd-jquery --save
```

Then what ?

Another question. Let's assume you can build the UI page independently from the server.
How would you deploy this page, containing CometD client-side code ?
You put the JavaScript files into a CDN ? Nginx ?
And then you deploy separately the CometD server-side Java code as a `war` file in a Servlet Container ?

Bottom line is that because of the way CometD binds to multiple toolkits, it will be a lot of work to maintain Bower support for all the toolkits.
I really need to be convinced that Bower is of tremendous help to develop CometD applications (I'm sure it's useful for other cases, but for CometD ?).

That's why I am asking for real cases where one would comment: "if I had Bower's cometd-jquery I would cut my development time by X% because I would save time in this, this and this steps that without Bower would require me a lot of time spent in boilerplate activities."

Thanks !

**Comment by lordkada** [comment]
@sbordet , don't worry... :-)

**I already ported CometD (jquery version) to be a bower package (because I needed) and I already use it in a real app**... (of course I haven't published it to the official bower repo because I'm not an official CometD maintainer and I don't want to mess; pls look at the third dependency in the bower I posted before:
```
"cometd-javascript-bower": "git@github.com:lordkada/cometd-javascript-bower.git#v3.0.5",
```

These applications are not static pages (the code is static, but also a compiled java class is a static file!), they are fully-featured and usually served through CDN... You don't need any servlet container or whatever to just run your UI... 

I cannot share my sources, but if it helps we could arrange an hangout to show you something...

**Comment by ffissore** [comment]
Hi

I've setup a travis-ci job at https://travis-ci.org/ffissore/cometd-bower-integration/builds.
It fetches this repo and publishes two bower packages: "cometd" and "cometd-binding-jquery". They are already available via `bower install`

The scripts run by the job are available at https://github.com/ffissore/cometd-bower-integration

The resulting bower compatible repositories are https://github.com/ffissore/bower-cometd and https://github.com/ffissore/bower-cometd-jquery

I would appreciate your feedback about the completeness of those repos (are all the needed files in place?). As I'm pretty new to cometd, any suggestion or contribution is welcome.

I'll be happy to give repos ownership to cometd, should @sbordet or other jetty cometd guys and gals ask for it.
The leitmotiv is to keep this repo (cometd) as "the only source of truth", and generating other repos contents, making tools like bower happy.
It requires a minimum amount of maintenance: editing .travis.yml and changing the [target tag](https://github.com/ffissore/cometd-bower-integration/blob/master/.travis.yml#L6). Travis builds are triggered as soon as a push occurs.

**Comment by ffissore** [comment]
I hope you had a chance to checkout if the two repos above contain the correct files. Can you help with dojo? Are files
```
dojox/cometd.js
dojox/cometd/*
```
all you need to bind cometd to dojo? Which version of dojo is used?

**Comment by sbordet** [comment]
CometD is now deployed to Bower.
As part of #649, there are now "vanilla" transports, i.e. not dependent on jQuery or Dojo, but using vanilla JavaScript APIs.
What's deployed in Bower are the vanilla transports only.
